qi-lib/flow/core/compiler.rkt

```diff
@@ -18,8 +18,53 @@
   (define (compile-flow stx)
     #`(qi0->racket #,(optimize-flow stx)))
 
+  (define (optimization-pass stx)
+    (syntax-parse stx
+      ;; restorative optimization for "all"
+      [((~datum ~>) ((~datum ><) onex) (~datum AND))
+       #`(esc (give (curry andmap #,(compile-flow #'onex))))]
+      ;; "deforestation" for values
+      [((~datum ~>) _0 ... ((~datum pass) f) ((~datum ><) g) _1 ...)
+       #'(~> _0 ... (>< (if f g ⏚)) _1 ...)]
+      [((~datum ~>) _0 ... ((~datum ><) g) ((~datum pass) f) _1 ...)
+       #'(~> _0 ... (>< (~> g (if f _ ⏚))) _1 ...)]
+      ;; merge amps in sequence
+      [((~datum ~>) _0 ... ((~datum ><) f) ((~datum ><) g) _1 ...)
+       #'(~> _0 ... (>< (~> f g)) _1 ...)] ; TODO: optimizing the inner flow?
+      ;; merge pass filters in sequence
+      [((~datum ~>) _0 ... ((~datum pass) f) ((~datum pass) g) _1 ...)
+       #'(~> _0 ... (pass (and f g)) _1 ...)]
+      ;; identities
+      [((~datum ~>) _0 ... ((~datum ~>) f ...) _1 ...)
+       #'(~> _0 ... f ... _1 ...)]
+      [((~datum ~>>) _0 ... ((~datum ~>>) f ...) _1 ...)
+       #'(~>> _0 ... f ... _1 ...)]
+      [((~datum ~>) _0 ... (~datum _) _1 ...)
+       #'(~> _0 ... _1 ...)]
+      [((~datum ~>>) _0 ... (~datum _) _1 ...)
+       #'(~>> _0 ... _1 ...)]
+      [((~datum ~>) (~datum _) ...)
+       #'_]
+      [((~datum ==) (~datum _) ...)
+       #'_]
+      [((~datum ><) (~datum _))
+       #'_]
+      [((~datum -<) f)
+       #'f]
+      [((~datum -<) _0 ... ((~datum gen) a ...) ((~datum gen) b ...) _1 ...)
+       #'(-< _0 ... (gen a ... b ...) _1 ...)]
+      [((~datum ~>) _0 ... (~datum △) (~datum ▽) _1 ...)
+       #'(~> _0 ... _1 ...)]
+      [((~datum ~>) _0 ... (~datum ▽) (~datum △) _1 ...)
+       #'(~> _0 ... _1 ...)]
+      ;; return syntax unchanged if there are no known optimizations
+      [_ stx]))
+
   (define (optimize-flow stx)
-    stx))
+    (let ([optimized (optimization-pass stx)])
+      (if (eq? optimized stx)
+          stx
+          (optimize-flow optimized)))))
```

Comment:
It would be nice to guarantee that this terminated 😅

Thought: maybe an upper bound on the number of loops is acceptable? The bound could be computed as a function of the "depth" of the flow expression.

Reply:
Good call. I don't know if this is the best way to structure the optimization passes in any case, but upper bound as a function of depth sounds like a good heuristic to keep in mind in case we end up doing something like this 👍